llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp

```diff
@@ -2008,8 +2008,10 @@ AMDGPULegalizerInfo::AMDGPULegalizerInfo(const GCNSubtarget &ST_,
 
   // S64 is only legal on SALU, and needs to be broken into 32-bit elements in
   // RegBankSelect.
-  auto &SextInReg = getActionDefinitionsBuilder(G_SEXT_INREG)
-    .legalFor({{S32}, {S64}});
+  auto &SextInReg =
+      getActionDefinitionsBuilder(G_SEXT_INREG)
+          .legalFor({{S32}, {S64}})
+          .widenScalarIf(typeIs(0, S16), widenScalarOrEltToNextPow2(0, 32));
```

Comment:
Probably ought to loosen this to clamp minimum to 32. Every other type should also round up to 32